### ISSUE TITLE
Problem with World of Xeen and Staging 0.81-RC

### ISSUE THREAD

**Issue by willow3100** [dosbox-staging/dosbox-staging#3277]
### Are you using the latest Dosbox-Staging Version?

- [X] I have checked releases and am using the latest release.

### What Operating System are you using?

Windows 11

### Relevant hardware info

ryzen 3900x, rtx2060,

### Have you checked that no other similar issue already exists?

- [X] I have searched and not found similar issues.

### A clear and concise description of what the bug is.

I have find a bug with dosbox staging 0.81RC. In inventory of World of xeen, cursor dispappeared even at 40 000 cycles. I must have 125 000 to see the cursor at the top. With dosbox staging 0.80.1, dosbox ece, dosbox vanilla, 40 000 is sufficient. 
dosbox staging 0.80.1 https://cdn.discordapp.com/attachments/626653390830698516/1189487658792144947/IMG_20231227_093553.jpg?ex=659e57b5&is=658be2b5&hm=7942d72080ab88f160f1936d9dd59670c1e40978b0c7e470cbf73c49401900cc&
dosbiox staging 0.81RC
https://cdn.discordapp.com/attachments/626653390830698516/1189487771182710895/IMG_20231227_093628.jpg?ex=659e57d0&is=658be2d0&hm=cce4a5c59b56f98bd4e464be2fcbebb27534133899688a04f5cd3b5d8545309f&
https://cdn.discordapp.com/attachments/626653390830698516/1189487882843471882/IMG_20231227_093641.jpg?ex=659e57ea&is=658be2ea&hm=c9176e2d1a0fb1d33846b8219395748661f8f21aa491f4b67ac64251e400325e&
https://cdn.discordapp.com/attachments/626653390830698516/1189487965492215878/IMG_20231227_093653.jpg?ex=659e57fe&is=658be2fe&hm=5a849432d26647ac8716cfaf63a8906a9f54d9885e72dc00d9bdb2d9a787be83&

### Steps to reproduce the behaviour.

Explain how to reproduce
1. launch dosbox and world of xeen
2. selct 125 000 cvycles and no problem
3.below 125 000 cursor begins to disappear or disappear

### Your configuration

```ini
[sdl]
priority=higher,normal
output=openglnb
fullresolution=1600x1200
windowresolution=1600x1200
mapperfile=mapper-ECE.map

[cpu]
core=dynamic
cycles=40000

[fluidsynth]
soundfont=FluidR3_GM_GS.sf2

[speaker]
pcspeaker=false
ps1audio=off

[dos]
ver=

[dosbox]
captures=..\captures\25

[render]
#glshader=crt-auto
aspect             = default
integer_scaling    = auto

[mixer]
rate=44100

[midi]
mididevice=fluidsynth
```

### Code of Conduct & Contributing Guidelines

- [X] Yes, I agree.

**Comment by johnnovak** [comment]
Will test on hardware today later, but before that here's are base truths about the game:

https://www.vogonswiki.com/index.php/List_of_CPU_speed_sensitive_games

**Game title**
Might & Magic: World of Xeen (1994) 

**Symptoms with faster CPUs**
Flashing mouse cursor when hovering over game window, map, and char page

**Maximum suitable CPU** 
486 40 

**Fix available?**
Reduce clock speed or disable cache with [Setmul](http://www.vogons.org/viewtopic.php?t=38613)

---

So the cursor issue is present on real hardware. 486/40 that roughly corresponds to about 15k cycles, so "fixing" the game with 30-40k cycles has always been a lucky coincidence at best. But we'll know more after my testing on hardware!

**Comment by johnnovak** [comment]
The game is perfectly playable with 0.81-RC1 with setting cycles to 7000. That's in line with the info from Vogons. The mouse pointer is a bit jerky, but that's life. I'll verify if it's jerky on hardware too.

The pointer starts disappearing when you go to 8k cycles or above.

```ini
[cpu]
cycles = 7000

[mouse]
mouse_raw_input = off

[autoexec]
mousectl -r 100
```

As shown, the pointer never disappears, but flickers a little on the top of the inventory screens at 7k cycles.

![image](https://github.com/dosbox-staging/dosbox-staging/assets/698770/2512bb85-afbe-4c42-9b46-9391a5263eba)

**Comment by johnnovak** [comment]
Test 2. If I increase cycles to around 50-60k, the cursor is visible again in the inventory screens. It gets a little bit cut off at the very top at around 52k cycles still as shown below, but who cares; there's nothing to click there anyway.

Note the cursor only does its strange cycles-dependent "disappearance act" in inventory screen. It's always fully visible in the game world screen, no matter the cycles setting.

![image](https://github.com/dosbox-staging/dosbox-staging/assets/698770/27875357-d7f9-46ad-84eb-ef39f53caf78)

**Comment by johnnovak** [comment]
Okay, I've tested this on my real hardware Pentium MMX 200 PC downclocked to 125 MHz. I've tested a few different speed scenarios:

1. At 125 MHz, which roughly corresponds to about **60-80k cycles.**
2. At roughly middle-of-the-road 486 speeds by using SETMUL at the slowest possible setting. This corresponds to the **15-30k cycles range** roughly.
3. At moderate 386 speeds by using SETMUL at the slowest possible setting AND disabling all hardware caches in the BIOS. This corresponds to about **3000 to 6000 cycles.**

With this many data points, the truth has nowhere to hide! 😎 Ready? Then let's begin!

---

## Test 1

At 125 MHz (60-80k) cycles, I'm getting almost the exact behaviour as in Staging up to about 40k cycles. The pointer moves _relatively_ slowly (it's still a bit crap, but it's not jumping wildly around), and only in the character screens, the cursor disappear in the upper half of the screen when you stop moving it around. When you move it, it flickers like crazy but always disappears if you stop moving in the upper half. Here's the video proof:

https://drive.google.com/file/d/1bD9NDsRHM69wQd0MYRTZidKOGMHHZa0r/view?usp=sharing

As the flicker is gone from 50k cycles with 0.81-RC, we're doing _better_ than the hardware for this particular buggy game, actualy. This is good.

## Test 2

Interestingly, 486 speeds did not help at all. This is roughly 15-30k cycles territory. I could not notice much of a difference compared to running at 125 Mhz, so no video.

## Test 3

By slowing things down to 386 speeds (about 3000 to 7000 cycles) I was able to get 90-95% the same behaviour as with Staging at 7000 cycles. The mouse movement became jerky as hell on the in-game screen but remained smooth in the menu screens, and the pointer never disappeared. Note how jerky the movement is in the in-game screens; I'm moving the mouse continuously and _smoothly_ on my desk, and this is the result.

https://drive.google.com/file/d/1rSTlWBlnTMaXhAAiXIeaLpxah6o_KCBi/view?usp=sharing

---

TL;DR the game is a buggy mess, and we're emulating the hardware behaviour accurately for better or worse. The recommended way to play the game is to use 7000 cycles and embrace the jerkiness -- that's what people had to endure back in the day on hardware too! 😅  

Alternatively, you can use 50-60k cycles to get less jerky mouse behaviour if you have a more powerful machine (emulating 60k cycles is still small potato on middle-of-the-road hardware from the last 8-10 years). 

We can't fix buggy games at the emulator level, and how other forks behave with less accurate emulation behaviour is irrelevant.

Not an emulator bug, case closed 😎

**Comment by rderooy** [comment]
eXoDOS v6 actually uses 30k cycles. With this in either DOSBox Staging or DOSBox-X, the mouse cursor flickers at the top of the screen, but otherwise seems to be fine.
```ini
# Config file for Dosbox-ECE r4230
[sdl]
fullscreen=false
fulldouble=false
fullresolution=0x0
windowresolution=1280x960
output=overlay
mapperfile=mapper.map
[dosbox]
machine=svga_s3
memsize=64
[render]
aspect=true
scaler=normal2x
[cpu]
core=auto
cputype=auto
cycles=30000
[pci]
[mixer]
[midi]
fluid.soundfont=.\mt32\SoundCanvas.sf2
mt32.romdir=.\mt32
[sblaster]
[gus]
[speaker]
[joystick]
[serial]
[dos]
[ipx]
[autoexec]
imgmount d .\eXoDOS\MM6\cd\NEW_WORLD_1_WORLD_OF_XEEN.cue .\eXoDOS\MM6\cd\NEW_WORLD_2_WORLD_OF_XEEN.cue -t cdrom
@mount c .\eXoDOS\MM6
c:
cls
run.bat
exit
```

**Comment by johnnovak** [comment]
> pdate: Running it with DOSBox SVN-ECE r4358 does not cause the mouse flicker.

Given we've established that flickering is _the_ accurate hardware behaviour, that makes ECE's approach to VGA rendering at the very least suspect.

**Comment by Grounded0** [comment]
If you want to emulate 60k cycles on a lower end or older hardware you could do something like this by forcing dynamic core on real mode:

`XEEN.BAT`:

```
config -set "core=dynamic"
config -set "cycles=auto 275000 50%%"
GAME.EXE
config -set "core=auto"
config -set "cycles=auto 12000 50%%"
```

**Comment by kcgen** [comment]
Fantastic to see this reproduced on hardware, right down to the time-pulsing jumps of the cursor, as well as the slight flickering of it in two sequential positions!

https://media.tenor.com/FCU8re-B8PcAAAAM/baby-scream-yeah.gif

Very nice work getting all these setup on hardware, @johnnovak !

**Comment by johnnovak** [comment]
Cheers @kcgen. We can chase our tails forever without establishing the base truths on real hardware first. I will add my recent findings to the list of problematic games in our wiki to preserve this info for posterity.

**Comment by willow3100** [comment]
> @kcgen @willow3100 @kklobe
> 
> Okay, I've tested this on my real hardware Pentium MMX 200 PC downclocked to 125 MHz. I've tested a few different speed scenarios:
> 
>     1. At 125 MHz, which roughly corresponds to about **60-80k cycles.**
> 
>     2. At roughly middle-of-the-road 486 speeds by using SETMUL at the slowest possible setting. This corresponds to the **15-30k cycles range** roughly.
> 
>     3. At moderate 386 speeds by using SETMUL at the slowest possible setting AND disabling all hardware caches in the BIOS. This corresponds to about **3000 to 6000 cycles.**
> 
> 
> With this many data points, the truth has nowhere to hide! 😎 Ready? Then let's begin!
> ## Test 1
> 
> At 125 MHz (60-80k) cycles, I'm getting almost the exact behaviour as in Staging up to about 40k cycles. The pointer moves _relatively_ slowly (it's still a bit crap, but it's not jumping wildly around), and only in the character screens, the cursor disappear in the upper half of the screen when you stop moving it around. When you move it, it flickers like crazy but always disappears if you stop moving in the upper half. Here's the video proof:
> 
> https://drive.google.com/file/d/1bD9NDsRHM69wQd0MYRTZidKOGMHHZa0r/view?usp=sharing
> 
> As the flicker is gone from 50k cycles with 0.81-RC, we're doing _better_ than the hardware for this particular buggy game, actualy. This is good.
> ## Test 2
> 
> Interestingly, 486 speeds did not help at all. This is roughly 15-30k cycles territory. I could not notice much of a difference compared to running at 125 Mhz, so no video.
> ## Test 3
> 
> By slowing things down to 386 speeds (about 3000 to 7000 cycles) I was able to get 90-95% the same behaviour as with Staging at 7000 cycles. The mouse movement became jerky as hell on the in-game screen but remained smooth in the menu screens, and the pointer never disappeared. Note how jerky the movement is in the in-game screens; I'm moving the mouse continuously and _smoothly_ on my desk, and this is the result.
> 
> https://drive.google.com/file/d/1rSTlWBlnTMaXhAAiXIeaLpxah6o_KCBi/view?usp=sharing
> 
> TL;DR the game is a buggy mess, and we're emulating the hardware behaviour accurately for better or worse. The recommended way to play the game is to use 7000 cycles and embrace the jerkiness -- that's what people had to endure back in the day on hardware too! 😅
> 
> Alternatively, you can use 50-60k cycles to get less jerky mouse behaviour if you have a more powerful machine (emulating 60k cycles is still small potato on middle-of-the-road hardware from the last 8-10 years).
> 
> We can't fix buggy games at the emulator level, and how other forks behave with less accurate emulation behaviour is irrelevant. More importantly, **it looks like there's no perfect CPU speed for the game on real hardware either.** You're just trading one compromise for another by varying the CPU speed.
> 
> So, this is not an emulator bug, and there's nothing to fix on our side. Case closed 😎
Thanks for your tests.

I will use 125 000 cycles and I hope some new bugs won't appeared.

**Comment by Grounded0** [comment]
We're stating that all cards based on the S3 Vision graphics core (86x/96x/Trio/ViRGE) likely shows this behaviour. Tseng ET4000 series including the ET6000 (which i had) may or may not perform this way or the ATI Mach 64 series. These 3 manufacturers were performance kings back in the day and were very popular. These would have to be verified against genuine hardware.

Main function of cards from this era was to perform as Windows accelerator cards by implementing routines from IBM XGA standard that crashed and burned and greatly improved performance. I'm seeing 20x performance increase in [SFPD Homicide / Case File: The Body in the Bay (1995)](https://www.mobygames.com/game/4061/sfpd-homicide-case-file-the-body-in-the-bay/) just by enabling WinG for the sometimes very sluggish game.

You can read more about video cards from this era from the very excellent [VGA Museum](https://www.vgamuseum.info/).

**Comment by johnnovak** [comment]
> Thanks for your tests. I had play many times might&magic 4 and might&magic 5 with my 386 and my pentium90 at this time but I don't remeber these problems. At that time, I had to be very tolerant, but I was a teenager.

I haven't tested it, but I've read that the earlier standalone MM4 and MM5 games are not prone to these problems, only Xeen. Maybe you played those standalone games, and that's what you remember.

> At the end, I will use 125 000 cycles with staging 0.81 and I hope some new bugs won't appeared because of this huge numbers of cycles.

Well, but no need for 125k. As I wrote, 50k fixes the problem, and 7k cycles is also alright, just the cursor is a bit janky but you kinda get used to it.

**Comment by rderooy** [comment]
I tried it in 86Box with a 486DX266 and a few graphics cards (ATI, ET3000 and ET4000), and I got flickering in all cases. But I have no idea how accurately that reflects real cards.

**Comment by willow3100** [comment]
> > 
> I haven't tested it, but I've read that the earlier standalone MM4 and MM5 games are not prone to these problems, only Xeen. Maybe you played those standalone games, and that's what you remember.
> 
I have play standalone MM4 because MM5 didn't existed ah this time (I have my floppy disc of MM4 and MM5 but my new Pc don't allow to use floppy disc drive). When MM5 was released, I have install world of Xeen (merge of 4 and 5) and play world of xeen and  not standalone MM5. 

I had a cirrus logic card with my pentium 90. With my 386sx20, I don't know. Maybe the problem didn't existed with this cirrus logic card or it's my memory that failed.
I didn't have anymore my 386 and my pentium 90. I can't test anymore but I have no memory of this problem even on my pentium 90. I am too old sic!!!

> > 
> Well, but no need for 125k. As I wrote, 50k fixes the problem, and 7k cycles is also alright, just the cursor is a bit janky but you kinda get used to it.
Yes. 125K is to see the cursor still the top of screen.

**Comment by johnnovak** [comment]
> I have play standalone MM4 because MM5 didn't existed ah this time (I have my floppy disc of MM4 and MM5 but my new Pc don't allow to use floppy disc drive). When MM5 was released, I have install world of Xeen (merge of 4 and 5) and play world of xeen and not standalone MM5.

Yeah, so to be clear, I've done the testing with the CD re-release of Xeen. Merging MM4 and MM5 manually would probably yield different (and better) results. If you have spare time, I'd be interesting to confirm that with merged MM4 & MM5 floppy versions the issue is not present.

> I had a cirrus logic card with my pentium 90. With my 386sx20, I don't know. Maybe the problem didn't existed with this cirrus logic card or it's my memory that failed. I didn't have anymore my 386 and my pentium 90. I can't test anymore but I have no memory of this problem even on my pentium 90. I am too old sic!!!

Sadly, we don't emulate the Cirrus Logic, that's an old VGA card, and many old demos and perhaps even games only work correctly on it. I used to have a Trident 9400 CXi VLB in my 486DX2/66 😄

For the record, I doubt the mouse issue is video card related.

**Comment by willow3100** [comment]
> > 
> Yeah, so to be clear, I've done the testing with the CD re-release of Xeen. Merging MM4 and MM5 manually would probably yield different (and better) results. If you have spare time, I'd be interesting to confirm that with merged MM4 & MM5 floppy versions the issue is not present.
> 
Problem is present with merged floppy french version because it's the version that I use. I don't know if problem exist with standalone floppy disc version because I don't have these versions on my hard disc.

> Sadly, we don't emulate the Cirrus Logic, that's an old VGA card, and many old demos and perhaps even games only work correctly on it. I used to have a Trident 9400 CXi VLB in my 486DX2/66 😄
> 
I know (I have never seen dosbox support of cirrus logic card even if cirrus logic was successfull at this time) but it would be interesting to know if the problem existed because I d'ont have any memory of these problem in 1995.

**Comment by rderooy** [comment]
I tried it in 86Box with a CL GD5426 and had the same flickering.
